### PR TITLE
3.1 - Jsonview auto serialize

### PR DESCRIPTION
**Pull request by ADmad**
Refs #6500

### CODE REVIEW

File: src/View/JsonView.php [hunk 1]
```diff
@@ -48,11 +45,13 @@
  *
  * `{"posts": [...], "users": [...]}`
  *
- * If you don't use the `_serialize` key, you will need a view. You can use extended
- * views to provide layout-like functionality.
+ * If you need the regular view template processing you can set `_serialize` to
+ * false which will turn off automatic serialization and instead view template
+ * and layotu will be used.
```
Comment (ceeram) [inline src/View/JsonView.php]: nitpick typo in layout

**Comment by ceeram** [comment]
lgtm

**Comment by ceeram** [comment]
Shouldnt XmlView behave similar?

**Comment by markstory** [comment]
XmlView should behave in a similar way. Having them be different will be not good. Also this change could be interpreted as an API breaking change. If an app was relying on using view files, it would now default to serializing which could be unexpected. What was wrong with the `serialize => true` approach?

**Comment by ADmad** [comment]
> What was wrong with the serialize => true approach?

It would still require adding `$this->set('_serialize', true)` in all actions which isn't DRY.

File: src/View/JsonView.php (hunk 1 above)
Comment (ADmad) [inline src/View/JsonView.php]: Not a nitpick, this ends up in API docs :)

File: src/View/JsonView.php [hunk 2]
```diff
@@ -22,22 +22,19 @@
...
  *
  * ```
- * $this->set(['posts' => $posts, '_serialize' => 'posts']);
+ * $this->set(['posts' => $posts]);
  * ```
  *
- * When the view is rendered, the `$posts` view variable will be serialized
- * into JSON.
+ * By default all view variables will be serialized to JSON. So for above example
+ * the `$posts` view variable will be serialized into JSON.
```
Comment (markstory) [inline src/View/JsonView.php]: Is this still true?

Comment (ADmad) [inline src/View/JsonView.php]: Yes it is.

**Comment by josegonzalez** [comment]
Seems like an API break, and perhaps a serious one for certain users. As such, I'm :-1: .

Do we want to move to breaking the API in minor releases?

**Comment by patrickconroy** [comment]
Would a controller property work? So, setting serialize in your AppController to true would auto-serialize the json view? And it'd default to false?

Right now all my json stuff has that serialize view property which is cumbersome.

**Comment by dereuromark** [comment]
@patrickconroy You can set `$this->set('_serialize', true)` in your AppController. This way it would not hurt if that controller doesn't use it, but it would apply to all that use JsonView by default.
That is pretty DRY in my book - and it is fully BC.

So that's what we should do here IMO.

**Comment by ADmad** [comment]
I'll switch back to not serializing by default to avoid BC issues. Instead can document the fact that one can set `$this->set('_serialize', true)` in `AppController::beforeRender()` to get serializing as default behavior.

**Comment by markstory** [comment]
> Instead can document the fact that one can set $this->set('_serialize', true) in AppController::beforeRender() to get serializing as default behavior.

This sounds like a much safer plan. We could even make this part of the AppController in the project skeleton.

**Comment by ADmad** [comment]
Done.

**Comment by dereuromark** [comment]
This would probably have fit into 3.0 now as well. But with 3.1 it can be communicated better.
Also we should make XmlView allow true here.

**Comment by ADmad** [comment]
> This would probably have fit into 3.0 now as well.

No it won't as this change causes [this](https://github.com/cakephp/cakephp/blob/3.0.3/tests/TestCase/View/JsonViewTest.php#L114) to fail.

**Comment by dereuromark** [comment]
Doesn't make much sense, as you said :) But all right, 3.1 is fine.

**Comment by ADmad** [comment]
- Implemented support for `_serialize` true for `XmlView` too.
- Pulled out duplicate code from `JsonView` and `XmlView` to `SerializedView`

File: src/View/XmlView.php [hunk 3]
```diff
@@ -73,60 +74,18 @@ class XmlView extends View
...
+     * @var string
      */
-    public function loadHelpers()
-    {
-        if (isset($this->viewVars['_serialize'])) {
-            return;
-        }
-        parent::loadHelpers();
-    }
+    public $_responseType = 'xml';
```
Comment (ceeram) [inline src/View/XmlView.php]: why underscore public attribute?

File: src/View/SerializedView.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ * Parent class for view classes generating serialized outputs like JsonView and XmlView.
+ */
+class SerializedView extends View
+{
+    /**
+     * Response type.
+     *
+     * @var string
+     */
+    public $_responseType;
```
Comment (ceeram) [inline src/View/SerializedView.php]: underscored public attribute here too

File: src/View/JsonView.php [hunk 5]
```diff
@@ -72,50 +75,32 @@ class JsonView extends View
...
-        EventManager $eventManager = null,
-        array $viewOptions = []
-    ) {
-        parent::__construct($request, $response, $eventManager, $viewOptions);
-
-        if ($response && $response instanceof Response) {
-            $response->type('json');
-        }
-    }
+    public $_responseType = 'json';
```
Comment (ceeram) [inline src/View/JsonView.php]: and here

File: src/View/XmlView.php (hunk 3 above)
Comment (ADmad) [inline src/View/XmlView.php]: Mistake in all 3 classes. Should be protected.

**Comment by markstory** [comment]
Looks good to me. This looks far safer than the original approach.

**Comment by patrickconroy** [comment]
Hey guys. With the way `_serialize=true` is working it doesn't seem possible to have it work when you have a single var.

```
$status = 1;
$this->set(compact('status') + ['_serialize' => true]);
```

If you do the above right now, you get this response, which isn't valid JSON:
```
1
```

However, if you have more than one var, it works as expected:
```
$status = 1;
$a = 'b';
$this->set(compact('status', 'a') + ['_serialize' => true]);
```

Returns:
```
{
    "status": 1,
    "a": "b"
}
```

The unexpectedness of that makes me think this is a bug, but it's possible that's as intended.

**Comment by markstory** [comment]
@patrickconroy Could you open a new issue with that?

**Comment by patrickconroy** [comment]
Sure np.

**Comment by markstory** [comment]
Thanks.

**Pull request merged**